lib/fs.js

```diff
@@ -236,18 +236,28 @@ function nullCheck(path, propName, throwError = true) {
   return err;
 }
 
-function validatePath(path, propName) {
+function validatePath(path, propName, assertFilename) {
   let err;
+  const isString = typeof path === 'string',
+    isU8Array = isUint8Array(path);
 
   if (propName === undefined) {
     propName = 'path';
   }
 
-  if (typeof path !== 'string' && !isUint8Array(path)) {
+  if (!isString && !isU8Array) {
     err = new errors.TypeError('ERR_INVALID_ARG_TYPE', propName,
                                ['string', 'Buffer', 'URL']);
   } else {
     err = nullCheck(path, propName, false);
+    if (assertFilename && err === undefined) {
+      const lastChar = path[path.length - 1];
+
+      if (isString && lastChar === '/')
+        err = new errors.Error('ERR_PATH_IS_DIRECTORY', path);
+      if (isU8Array && lastChar === 47)
+        err = new errors.Error('ERR_PATH_IS_DIRECTORY', path);
+    }
   }
```

Comment:
Checking for an UInt8Array is not cheap. Therefore it would be good to rewrite this to:

```js
if (typeof path === 'string') {
  err = nullCheck(path, propName, false);
  if (assertFilename && err === undefined && path[path.length - 1] === '/')
    err = new errors.Errors(...);
} else if (isUint8Array(path)) {
  err = nullCheck(path, propName, false);
  if (assertFilename && err === undefined && path[path.length - 1] === 47)
    err = new errors.Error(...);
} else {
  err = new errors.TypeError(...);
}
```